src/dataflow/src/server.rs

```diff
@@ -1035,6 +1038,86 @@ where
+                            None => {
+                                // Clean up the expired weak handle.
+                                self.storage_state.persistence_frontiers.remove(&id);
+                                continue;
+                            }
+                        };
+
+                        let compaction_handle = match self.storage_state.compaction_handles.get(&id)
+                        {
+                            Some(handle) => handle,
+                            None => {
+                                continue;
+                            }
+                        };
+
+                        let compaction_handle = match compaction_handle.upgrade() {
+                            Some(handle) => handle,
+                            None => {
+                                // Clean up the expired weak handle.
+                                self.storage_state.compaction_handles.remove(&id);
+                                continue;
+                            }
+                        };
+
+                        // NOTE: It's **very** important that we don't allow compaction right up to
+                        // the persisted frontier. Otherwise, we won't be able to determine which
+                        // updates are valid and which are beyond the seal frontier when
+                        // restarting.
+                        let mut persistence_frontier = frontier_probe.with_frontier(|frontier| {
+                            Antichain::from(
+                                frontier
+                                    .iter()
+                                    .cloned()
+                                    .map(|e| e.saturating_sub(1))
+                                    .collect::<Vec<_>>(),
+                            )
+                        });
+
+                        println!(
+                            "persistence frontier for {}: {:?}, allowed_compaction: {:?}",
+                            id, persistence_frontier, frontier
+                        );
+
+                        persistence_frontier.meet_assign(&frontier);
+
+                        println!(
+                            "combined compaction frontier for {}: {:?}",
+                            id, persistence_frontier,
+                        );
+
+                        let id_frontiers = compaction_handle
+                            .1
+                            .iter()
+                            .map(|id| -> Result<_, _> {
+                                Ok((id.clone()?, persistence_frontier.clone()))
+                            })
+                            .collect::<Result<Vec<_>, persist::error::Error>>();
+
+                        match id_frontiers {
+                            Ok(id_frontiers) => {
+                                // TODO: This is blocking the worker!
+                                match compaction_handle.0.allow_compaction(&id_frontiers).recv() {
```

Comment:
driving by while doing other stuff but I was imagining that we would just fire off this allow_compaction and ignore the error? the compaction stuff is not strongly coupled between control and storage, and the `AllowCompaction` signals are more like a suggestion than a hard requirement? In the coordinator for example, we just fire off an `allow_compaction` request for tables and then spawn a new future to await the response and log iff theres an error

Reply:
I was envisioning sth like this, yes. 👌